lib/xnvme_be_linux_async_libaio.c

```diff
@@ -57,17 +74,41 @@ int
 _linux_libaio_poke(struct xnvme_queue *q, uint32_t max)
 {
 	struct xnvme_queue_libaio *queue = (void *)q;
-	struct timespec timeout = {.tv_sec = 0, .tv_nsec = 100000};
-	int min = queue->poll_io ? 0 : 1;
-	int completed = 0;
-
+	struct timespec timeout;
+	int min, completed;
 	max = max ? max : queue->base.outstanding;
 	max = max > queue->base.outstanding ? queue->base.outstanding : max;
 
-	completed = io_getevents(queue->aio_ctx, min, max, queue->aio_events, &timeout);
-	if (completed < 0) {
-		XNVME_DEBUG("FAILED: completed: %d, errno: %d", completed, errno);
-		return completed;
+	struct xnvme_aio_ring *ring = (struct xnvme_aio_ring *)queue->aio_ctx;
+
+	/* If ring is incompatible use io_getevents */
+	if (ring->magic != XNVME_AIO_RING_MAGIC || ring->incompat_features != 0) {
+		timeout.tv_sec = 0;
+		timeout.tv_nsec = 100000;
+		min = queue->poll_io ? 0 : 1;
+		completed = io_getevents(queue->aio_ctx, min, max, queue->aio_events, &timeout);
+		if (completed < 0) {
+			XNVME_DEBUG("FAILED: completed: %d, errno: %d", completed, errno);
+			return completed;
+		}
+
+		/* Otherwise copy events from memory */
+	} else {
+		uint32_t current = ring->head;
```

Comment:
Do you need to use an atomic load with acquire semantics here?

Reply:
> Do you need to use an atomic load with acquire semantics here?

Hmm, I guess it is not required, since worst case then we read an "older" ``head`` and consequently process less I/O. However, it should not lead to erroneous behavior? Or, am I missing here?

Reply:
Like, would there be a re-ordering issue her?

Reply:
No, I think you are right, sorry. The store release should be enough.